vitrivr-engine-module-pgvector/src/main/kotlin/org/vitrivr/engine/database/pgvector/descriptor/PgDescriptorInitializer.kt

```diff
@@ -0,0 +1,142 @@
+
+    /** The [Descriptor] prototype for this [PgDescriptorWriter]. */
+    protected val prototype = this.field.analyser.prototype(this.field)
+
+    /**
+     * Initializes the PostgreSQL table entity backing this [PgDescriptorInitializer].
+     */
+    override fun initialize() {
+        val statement = StringBuilder("CREATE TABLE IF NOT EXISTS \"${tableName}\" (")
+        statement.append("$DESCRIPTOR_ID_COLUMN_NAME uuid NOT NULL, ")
+        statement.append("$RETRIEVABLE_ID_COLUMN_NAME uuid NOT NULL, ")
+
+        /* Add columns for each field in the struct. */
+        for (field in this.prototype.schema()) {
+            when (field.type) {
+                Type.String -> statement.append("\"${field.name}\" varchar(255), ")
+                Type.Text -> statement.append("\"${field.name}\" text, ")
+                Type.Boolean -> statement.append("\"${field.name}\" boolean, ")
+                Type.Byte -> statement.append("\"${field.name}\" smallint, ")
+                Type.Short -> statement.append("\"${field.name}\" smallint, ")
+                Type.Int -> statement.append("\"${field.name}\" integer, ")
+                Type.Long -> statement.append("\"${field.name}\" bigint, ")
+                Type.Float -> statement.append("\"${field.name}\" real, ")
+                Type.Double -> statement.append("\"${field.name}\" double precision, ")
+                Type.Datetime -> statement.append("\"${field.name}\" datetime, ")
+                is Type.BooleanVector -> statement.append("\"${field.name}\" bit(${field.type.dimensions}), ")
+                is Type.DoubleVector -> statement.append("\"${field.name}\" vector(${field.type.dimensions}), ")
+                is Type.FloatVector -> statement.append("\"${field.name}\" vector(${field.type.dimensions}), ")
+                is Type.IntVector -> statement.append("\"${field.name}\" vector(${field.type.dimensions}), ")
+                is Type.LongVector -> statement.append("\"${field.name}\" vector(${field.type.dimensions}), ")
+            }
+        }
+
+        /* Finalize statement*/
+        statement.append("PRIMARY KEY ($DESCRIPTOR_ID_COLUMN_NAME), ")
+        statement.append("FOREIGN KEY ($RETRIEVABLE_ID_COLUMN_NAME) REFERENCES $RETRIEVABLE_ENTITY_NAME($RETRIEVABLE_ID_COLUMN_NAME) ON DELETE CASCADE);")
+
+        /* Create entity. */
+        try {
+            this.connection.jdbc.prepareStatement(/* sql = postgres */ statement.toString()).use {
+                it.execute()
+            }
+        } catch (e: SQLException) {
+            LOGGER.error(e) { "Failed to initialize entity '$tableName' due to exception." }
+            throw e
+        }
+
+        /* Create indexes (optional). */
+        for (index in this.field.indexes) {
+            try {
+                val indexStatement = when (index.type) {
+                    IndexType.SCALAR -> "CREATE INDEX ON $tableName (${index.attributes.joinToString(",")});"
```

Comment:
Do we not want to be able to specify what kind of index?

Reply:
I'm not sure that there are options in this case?

Reply:
There are several; see https://www.postgresql.org/docs/current/indexes-types.html. While BTREE as a default makes sense, some instances might also reasonably use HASH or BRIN. HASH, for example, can make sense for UUIDs, since there are no range queries on those anyway.

Reply:
I just checked, and the HASH index is known to be quite a bit faster if all you ever query for is equality, especially for large tables.

Reply:
Done!